# The Used Car Ads of Kleinanzeigen

This project uses a dataset of classified ads for used cars in Germany. The data requires clearning before exploration and analysis.

In [523]:
import pandas as pd
import numpy as np
pd.options.display.max_columns = None

In [524]:
autos = pd.read_csv('autos.csv', encoding = "Windows-1252")

In [525]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

In [526]:
autos.head(5)

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

We have a mix of data types here, mainly strings with some integers. There is null data in some cells. There are mixed dtypes in columns and the name label has additional information as part of the title. For example: Peugeot_807_160_NAVTECH_ON_BOARD has the make and model and information about having a navigation system. There is also an opportunity to translate some of the data into English or another language. 
There are 50,000 total entries in 20 columns.


In [527]:
print(autos.columns)

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


In [528]:
# selection note: df.loc[row_label, column_label] 
# df.loc[:,["col1","col2"]]
#rank = f500["rank"]
#rank_desc = rank.describe()

#prev_rank = f500["previous_rank"]
#prev_rank_desc = prev_rank.describe()

In [529]:
# convert column labels from camelcase to snakecase
column_edits = autos.rename(columns={
                            "yearOfRegistration": "registration_year", 
                            "monthOfRegistration":"registration_month",
                            "notRepairedDamage":"unrepaired_damage",
                            "dateCreated":"ad_created", 
                            "vehicleType": "vehicle_type",
                            "powerPS":"power_ps",
                            "fuelType":"fuel_type",
                            "nrOfPictures":"number_of_photos",
                            "postalCode":"postal_code",
                            "lastSeen":"last_viewed"
                            },
                    inplace = True
                    )

In [530]:
column_edits = autos.rename(columns={"offerType":"offer_type",
                                     "dateCrawled":"view_date"},
                            inplace = True)

In [531]:
column_edits = autos.rename(columns={"odometer":"odometer_km"},
                            inplace = True)

In [532]:
autos.head(5)

view_date                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model odometer_km  registration_month fuel_type  \
0    manuell       158  andere   150,000km                   3       lpg   
1  automatik       286     7er   150,000km                   6    benzin   
2    manuell       102    golf    70,000km                   7    benzin   
3  automatik        71  fortwo    70,000km                   6    benzin   
4    manuell         0   focus   150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  number_of_photos  \
0     peugeot              nein  2016-03-26 00:00:00                 0   
1         bmw              nein  2016-04-04 00:00:00                 0   
2  volkswagen              nein  2016-03-26 00:00:00                 0   
3       smart              nein  2016-03-12 00:00:00                 0   
4        ford              nein  2016-04-01 00:00:00                 0   

   postal_code          last_viewed  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

Let's look for anything that needs to be cleaned up. 

In [533]:
autos.describe(include = "all")

view_date         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-04-04 16:40:33  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

       odometer_km  registration_month fuel_type       brand  \
count        50000        50000.000000     45518       50000   
unique          13                 NaN         7          40   
top      150,000km                 NaN    benzin  volkswagen   
freq         32424                 NaN     30107       10687   
mean           NaN            5.723360       NaN         NaN   
std            NaN            3.711984       NaN         NaN   
min            NaN            0.000000       NaN         NaN   
25%            NaN            3.000000       NaN         NaN   
50%            NaN            6.000000       NaN         NaN   
75%            NaN            9.000000       NaN         NaN   
max            NaN           12.000000       NaN         NaN   

       unrepaired_damage           ad_created  number_of_photos   postal_code  \
count              40171                50000           50000.0  50000.000000   
unique                 2                   76               NaN           NaN   
top                 nein  2016-04-03 00:00:00               NaN           NaN   
freq               35232                 1946               NaN           NaN   
mean                 NaN                  NaN               0.0  50813.627300   
std                  NaN                  NaN               0.0  25779.747957   
min                  NaN                  NaN               0.0   1067.000000   
25%                  NaN                  NaN               0.0  30451.000000   
50%                  NaN                  NaN               0.0  49577.000000   
75%                  NaN                  NaN               0.0  71540.000000   
max                  NaN                  NaN               0.0  99998.000000   

                last_viewed  
count                 50000  
unique                39481  
top     2016-04-07 06:17:27  
freq                      8  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

In [534]:
autos["registration_year"].value_counts().head(20)

2000    3354
2005    3015
1999    3000
2004    2737
2003    2727
2006    2708
2001    2703
2002    2533
1998    2453
2007    2304
2008    2231
2009    2098
1997    2028
2011    1634
2010    1597
2017    1453
1996    1444
2012    1323
2016    1316
1995    1313
Name: registration_year, dtype: int64

In [535]:
autos["number_of_photos"].value_counts()

0    50000
Name: number_of_photos, dtype: int64

In [536]:
autos["postal_code"].value_counts().head(20)

10115    109
65428    104
66333     54
45888     50
44145     48
48599     47
65933     45
65719     44
15344     43
37154     42
50354     42
52525     42
38518     40
44339     40
32791     40
51065     40
77933     40
45881     40
30419     40
25524     39
Name: postal_code, dtype: int64

In [537]:
autos["price"].unique()

array(['$5,000', '$8,500', '$8,990', ..., '$385', '$22,200', '$16,995'],
      dtype=object)

In [538]:
autos["odometer_km"].value_counts()

150,000km    32424
125,000km     5170
100,000km     2169
90,000km      1757
80,000km      1436
70,000km      1230
60,000km      1164
50,000km      1027
5,000km        967
40,000km       819
30,000km       789
20,000km       784
10,000km       264
Name: odometer_km, dtype: int64

Let's edit the odometer and price columns and remove the text to convert to numerical values. There aren't decimals so we can use int.

In [539]:
autos["price"] = (autos["price"]
                  .str.replace("$","")
                  .str.replace(",","")
                  .astype(int)
                 )

In [540]:
autos = autos[autos["price"].between(1,350000)]
autos["price"].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64

The price and odometer columns are now numerical values. It looks like there are 1,421 entries under price for €0. Does this mean they were free? This could be an error. We should check on this later. We could also convert km to miles. 150,000 km is ~93,206. Cars over 100k miles in the US sell for much less. Volkswagens are also notorious for having expensive electrical issues after 80,000 miles. We could also look into how many Volkswagens are sold around this mileage!

In [541]:
autos["odometer_km"] = (autos["odometer_km"]
                        .str.replace("km","")
                        .str.replace(",","")
                        .astype(int))

In [542]:
autos["odometer_km"].value_counts().head(10)

150000    31414
125000     5057
100000     2115
90000      1734
80000      1415
70000      1217
60000      1155
50000      1012
5000        836
40000       815
Name: odometer_km, dtype: int64

In [543]:
autos["odometer_km"].value_counts().tail(10)

90000    1734
80000    1415
70000    1217
60000    1155
50000    1012
5000      836
40000     815
30000     780
20000     762
10000     253
Name: odometer_km, dtype: int64

Are there any odd values in either odometer or price columns?

In [544]:
prices = autos["price"]
prices.unique().shape

(2346,)

Above we see that max value of 99 million euros! Are people really buying cars over 100,000 euros? Let's sort and find out.

In [545]:
autos[autos["price"] > 100000].sort_values("price")

view_date                                               name  \
29286  2016-03-22 15:58:56  Porsche_911_Turbo_PDK__Sport_Chrono_Garantie_D...   
16964  2016-04-01 16:46:18                    Bentley_Continental_Supersports   
17540  2016-03-30 00:49:49   Porsche_930_Turbo_classic_analytics_Gutachten_2+   
49391  2016-03-18 00:55:16  Lamborghini_Gallardo_LP560_4_E_Gear_"Callisto_...   
22060  2016-03-09 00:58:24  Tesla_Model_X90D_Autopilot_Leder_AHK_Kaltwette...   
7402   2016-03-22 19:48:09  Porsche_911_Carrera_4S_Cabrio_PDK__BOSE__NEU__...   
21783  2016-03-26 22:06:24               Porsche_911_Carrera_4S_Cabriolet_PDK   
33884  2016-03-26 21:55:12               Porsche_911_Carrera_4S_Cabriolet_PDK   
38814  2016-03-22 16:53:44  Porsche_Porsche_964_TURBO_S_X33_WLS_ab_Werk_35...   
43282  2016-04-03 18:42:28               Porsche_911_Carrera_4S_Cabriolet_PDK   
2751   2016-03-15 10:52:35                               Porsche_911___993_4S   
44406  2016-03-15 14:47:26  Mercedes_Benz_170_S_Cabriolet_A__Startklar_fue...   
8232   2016-04-01 21:50:47               Porsche_993_S_Schalter_BRD_neuwertig   
1878   2016-03-12 16:58:37                                  Porsche_911_Turbo   
49815  2016-03-08 10:06:22  SUCHE_TIPPS___Ford_Mustang_Shelby_GT_350_500_K...   
14268  2016-04-01 12:52:27   Porsche_997_GT3_Approved_Lift_Unfallfrei_Mj.2011   
2454   2016-03-21 22:51:29                                    Porsche_911_GT3   
32185  2016-03-25 12:42:09  Porsche_997_GT3_MKII_EIN_SAMMLERSTÜCK_IN_VOLLA...   
49668  2016-03-31 13:45:02         Porsche_911__993__Carrera_S_"SONDERMODELL"   
33638  2016-03-24 12:54:19                                 Porsche_911_GT3_RS   
10500  2016-03-17 12:56:38                                        Porsche_991   
45387  2016-03-18 16:41:22  Porsche_911_GT3__Clubsp._Lift_nur_bis_25.03._v...   
22673  2016-03-05 18:40:13  Porsche_911_991_GT3_LED_Licht_Chrono_Navi_Club...   
18509  2016-03-27 18:43:02                                    Porsche_991_GT3   
32840  2016-03-22 08:55:27  Mercedes_Benz_G_500_7G_TRONIC_Cabrio_Desingno_...   
11433  2016-03-30 21:44:21  Audi_R8_V10_plus_5.2_FSI_quattro_S_tronic_Akra...   
17140  2016-03-05 17:39:22  Mercedes_Benz_SLS_AMG_Roadster_AMG_SPEEDSHIFT_DCT   
20351  2016-03-22 17:47:59                                    BMW_Z8_roadster   
28090  2016-03-08 19:50:06  Tesla_Model_X_P90D_Signature_Sondermodel__Neuw...   
43668  2016-03-16 18:47:26  Porsche_993/911_Turbo_WLS_II_Exclusive_S_deuts...   
40918  2016-03-20 18:40:05                             Porsche_911_991_GT3_RS   
37840  2016-03-21 10:50:12                                        Porsche_997   
38299  2016-03-28 22:25:25                                Glas_BMW_mit_Wasser   
47337  2016-04-05 10:25:38                                    BMW_Z8_roadster   
12682  2016-03-28 22:48:01  Porsche_GT3_RS__PCCB__Lift___grosser_Exklusiv_...   
35923  2016-04-03 07:56:23  Porsche_911_Targa_Exclusive_Edition__1_von_15_...   
34723  2016-03-23 16:37:29  Porsche_Porsche_911/930_Turbo_3.0__deutsche_Au...   
14715  2016-03-30 08:37:24                 Rolls_Royce_Phantom_Drophead_Coupe   
36818  2016-03-27 18:37:37                                        Porsche_991   

       seller offer_type   price   abtest vehicle_type  registration_year  \
29286  privat    Angebot  104900  control        coupe               2011   
16964  privat    Angebot  105000  control        coupe               2010   
17540  privat    Angebot  105000     test        coupe               1988   
49391  privat    Angebot  109999     test        coupe               2008   
22060  privat    Angebot  114400     test          suv               2016   
7402   privat    Angebot  115000     test       cabrio               2016   
21783  privat    Angebot  115991     test       cabrio               2013   
33884  privat    Angebot  116000  control       cabrio               2013   
38814  privat    Angebot  119500     test        coupe               1992   
43282 

In the above there are some sketchy entries. We have 12345678 and 11111111. These don't seem likely. The 1999 BMW 530d wagon's cost is listed as 1,234,566. This is definitely wrong for a car that's maybe worth $300. The limousine entries seem suspicious as well. There are 2 entries for Ferraris that are over 1 million euros (1,300,000 and 3,890,000) and a Maserati (10,000,000) - these seem like they could be reasonable prices so we will keep them.

In [546]:
prices.describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64

In [547]:
prices.value_counts().head(10)

500     781
1500    734
2500    643
1200    639
1000    639
600     531
3500    498
800     498
2000    460
999     434
Name: price, dtype: int64

In [548]:
prices.sort_index(ascending=False).head(10)

49999     1250
49998    22900
49997    13200
49996     1980
49995    24900
49994     5000
49993     1650
49992     4800
49991      500
49990    17500
Name: price, dtype: int64

In [549]:
autos["price"].dtypes

dtype('int64')

There are some good candidates for cleaning up. The odometer column could be edited to change column label to odometer_km and removing the "km" from the values. Price also has this issue, the dollar sign can be removed. There is also an odd value under price "$99,999,999." A 100 million deutchmark car was probably not sold over German ebay. Registration_year also has some issues. This would be a good place to replace any cars listed with registration_year as less than 1950 with Nan. Seller, offer type and number of photos can be deleted as they don't appear to have useful information.

In [550]:
autos["registration_year"].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

The registration years are from 1960 to 2006. So we can edit these.

In [551]:
autos = autos[autos["registration_year"].between(1900,2016)]

In [552]:
# check to make sure the range set makes sense
autos["registration_year"].describe()

count    46681.000000
mean      2002.910756
std          7.185103
min       1910.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64

In [553]:
# remove columns that only have 2 or fewer unique values
autos.drop(['seller','offer_type', 'number_of_photos'], axis = 1)

view_date                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
30     2016-03-14 11:47:31                         Peugeot_206_Unfallfahrzeug   
31     2016-03-14 16:53:09                          Noch_gut_erhaltenen_C_320   
...                    ...                                                ...   
49968  2016-04-01 17:49:15                  Mercedes_Benz_190_D_2.5_Automatik   
49969  2016-03-17 18:49:02             Nissan_X_Trail_2.2_dCi_4x4_Sport_m.AHZ   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_2.

Let's look at some more columns that seemed odd.

In [554]:
autos["vehicle_type"].value_counts(dropna=False)

limousine     12598
kleinwagen    10585
kombi          8930
bus            4031
cabrio         3016
NaN            2704
coupe          2462
suv            1965
andere          390
Name: vehicle_type, dtype: int64

In [555]:
autos["brand"].value_counts(dropna=False)

volkswagen        9862
bmw               5137
opel              5022
mercedes_benz     4503
audi              4041
ford              3263
renault           2201
peugeot           1393
fiat              1197
seat               853
skoda              766
nissan             713
mazda              709
smart              661
citroen            654
toyota             593
hyundai            468
sonstige_autos     458
volvo              427
mini               409
mitsubishi         384
honda              366
kia                330
alfa_romeo         310
porsche            286
suzuki             277
chevrolet          266
chrysler           164
dacia              123
daihatsu           117
jeep               106
subaru             100
land_rover          98
saab                77
jaguar              73
daewoo              70
trabant             65
rover               62
lancia              50
lada                27
Name: brand, dtype: int64

In [556]:
autos["unrepaired_damage"].value_counts(dropna=False)

nein    33834
NaN      8307
ja       4540
Name: unrepaired_damage, dtype: int64

We can translate a few categories here in vehicle_type. Kleinwagen means compact car. Combi means station wagon (could also be a hatchback?) Cabrio is a convertible. And andere just means other. Let's update these to English.

In [557]:
transl_dict = {
    "limousine": "limousine",
    "kleinwagen": "compact",
    "kombi": "station wagon",
    "bus": "bus",
    "cabrio": "convertible",
    "coupe": "coupe",
    "suv": "suv",
    "andere": "other"
    }

autos["vehicle_type"] = autos["vehicle_type"].map(transl_dict)

In [568]:
# check translations
print(autos["vehicle_type"].head(20))

0               bus
1         limousine
2         limousine
3           compact
4     station wagon
5               bus
6         limousine
7         limousine
8               NaN
9               bus
11              NaN
12          compact
13        limousine
14          compact
15            coupe
16    station wagon
17            coupe
18              bus
19              suv
20    station wagon
Name: vehicle_type, dtype: object


Calculate the distribution of values for date info.

In [559]:
print(autos["view_date"].str[:10].head(10))

0    2016-03-26
1    2016-04-04
2    2016-03-26
3    2016-03-12
4    2016-04-01
5    2016-03-21
6    2016-03-20
7    2016-03-16
8    2016-03-22
9    2016-03-16
Name: view_date, dtype: object


In [560]:
autos[["view_date","ad_created","last_viewed"]][0:5]

view_date           ad_created          last_viewed
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [561]:
print(autos["ad_created"].str[:10].head(10))

0    2016-03-26
1    2016-04-04
2    2016-03-26
3    2016-03-12
4    2016-04-01
5    2016-03-21
6    2016-03-20
7    2016-03-16
8    2016-03-22
9    2016-03-16
Name: ad_created, dtype: object


In [562]:
print(autos["last_viewed"].str[:10].head(10))

0    2016-04-06
1    2016-04-06
2    2016-04-06
3    2016-03-15
4    2016-04-01
5    2016-04-06
6    2016-03-23
7    2016-04-07
8    2016-03-26
9    2016-04-06
Name: last_viewed, dtype: object


We could really use a sold on date! It would be great to be able to determine average time from ad placed to car sold or ad removed. Also is price price-asked or price-sold-at?

In [563]:
autos["last_viewed"].value_counts().head(10)

2016-04-07 06:17:27    8
2016-04-07 03:16:17    7
2016-04-06 21:17:51    7
2016-04-07 05:16:17    6
2016-04-06 02:17:26    6
2016-04-05 16:44:47    6
2016-04-06 06:17:24    6
2016-04-07 06:46:12    6
2016-04-07 04:46:51    6
2016-04-07 03:45:23    6
Name: last_viewed, dtype: int64

In [564]:
#??????? 

autos["ad_created"].value_counts(normalize=True, dropna=False).head(10)

2016-04-03 00:00:00    0.039009
2016-03-20 00:00:00    0.038067
2016-03-21 00:00:00    0.037531
2016-04-04 00:00:00    0.036953
2016-03-12 00:00:00    0.036653
2016-04-02 00:00:00    0.035196
2016-03-07 00:00:00    0.035004
2016-03-14 00:00:00    0.035004
2016-03-28 00:00:00    0.034725
2016-03-15 00:00:00    0.034168
Name: ad_created, dtype: float64

In [598]:
brands = autos["brand"].value_counts(normalize = True) 

# find the top 15 brands
top_onefive_brands = brands[brands > 0.013].index
print(top_onefive_brands) 

# check for 15
len(top_onefive_brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat', 'skoda', 'nissan', 'mazda', 'smart',
       'citroen'],
      dtype='object')


15

In [599]:
avg_sale_per_brand = {} 

for brand in top_onefive_brands: 
    subset_brand = autos[autos["brand"] == brand] 
    avg_price = subset_brand["price"].mean() 
    avg_sale_per_brand[brand] = int(avg_price)
print(avg_sale_per_brand)

{'bmw': 8332, 'seat': 4397, 'volkswagen': 5402, 'ford': 3749, 'renault': 2474, 'opel': 2975, 'smart': 3580, 'nissan': 4743, 'mazda': 4112, 'audi': 9336, 'skoda': 6368, 'citroen': 3779, 'fiat': 2813, 'peugeot': 3094, 'mercedes_benz': 8628}


## Calculate the average mileage and average price of the top 15 

Now let's use the loop method to calculate the mean mileage and mean price for each of the top brands, storing the results in a dictionary.
- Convert both dictionaries to series objects, using the series constructor.
- Create a dataframe from the first series object using the dataframe constructor.
- Assign the other series as a new column in this dataframe.
- Print the dataframe and look at the aggregate data.

In [600]:
# from lesson 8:

bmp_series = pd.Series(brand_mean_prices)
print(bmp_series)

df = pd.DataFrame(bmp_series, columns=['mean_price'])
df

audi             9336
bmw              8332
citroen          3779
fiat             2813
ford             3749
mazda            4112
mercedes_benz    8628
nissan           4743
opel             2975
peugeot          3094
renault          2474
seat             4397
skoda            6368
smart            3580
volkswagen       5402
dtype: int64


mean_price
audi                 9336
bmw                  8332
citroen              3779
fiat                 2813
ford                 3749
mazda                4112
mercedes_benz        8628
nissan               4743
opel                 2975
peugeot              3094
renault              2474
seat                 4397
skoda                6368
smart                3580
volkswagen           5402

In [605]:
# new dictionary for mileage 

brand_avg_km = {}

for brand in avg_sale_per_brand:
    brand_only = autos[autos["brand"] == brand]
    avg_km = brand_only["odometer_km"].mean()
    brand_avg_km[brand] = int(avg_km)

In [606]:
brand_avg_km

{'audi': 129157,
 'bmw': 132572,
 'citroen': 119694,
 'fiat': 117121,
 'ford': 124266,
 'mazda': 124464,
 'mercedes_benz': 130788,
 'nissan': 118330,
 'opel': 129310,
 'peugeot': 127153,
 'renault': 128071,
 'seat': 121131,
 'skoda': 110848,
 'smart': 99326,
 'volkswagen': 128707}

In [608]:
avg_km = pd.Series(brand_avg_km).sort_values(ascending=False)
avg_km

bmw              132572
mercedes_benz    130788
opel             129310
audi             129157
volkswagen       128707
renault          128071
peugeot          127153
mazda            124464
ford             124266
seat             121131
citroen          119694
nissan           118330
fiat             117121
skoda            110848
smart             99326
dtype: int64

In [609]:
avg_prices = pd.Series(avg_sale_per_brand).sort_values(ascending=False)
avg_prices

audi             9336
mercedes_benz    8628
bmw              8332
skoda            6368
volkswagen       5402
nissan           4743
seat             4397
mazda            4112
citroen          3779
ford             3749
smart            3580
peugeot          3094
opel             2975
fiat             2813
renault          2474
dtype: int64

In [610]:
nice_chart = pd.DataFrame(avg_km,columns=['avg_kilometers'])
nice_chart

mean_kilometers
bmw                     132572
mercedes_benz           130788
opel                    129310
audi                    129157
volkswagen              128707
renault                 128071
peugeot                 127153
mazda                   124464
ford                    124266
seat                    121131
citroen                 119694
nissan                  118330
fiat                    117121
skoda                   110848
smart                    99326

In [614]:
nice_chart["avg_price"] = avg_prices
nice_chart

mean_kilometers  mean_price
bmw                     132572        8332
mercedes_benz           130788        8628
opel                    129310        2975
audi                    129157        9336
volkswagen              128707        5402
renault                 128071        2474
peugeot                 127153        3094
mazda                   124464        4112
ford                    124266        3749
seat                    121131        4397
citroen                 119694        3779
nissan                  118330        4743
fiat                    117121        2813
skoda                   110848        6368
smart                    99326        3580

## Conclusion 

All right! This wraps up this project. I got stuck for a bit on the average km calculations. I had an error when I was editing the prices category. Once that was solved things went smoother. It looks like Germans have a preference for purchasing used German cars. The most popular cars sold are all German companies. In the top 15 40% are German cars and 33% are from the European Union. American cars don't apper to be favored much, not surprise there!